odl/operator/fspace_ops.py

```diff
@@ -0,0 +1,412 @@
+        slice2[axis] = 1
+        out[slice1] = f_data[slice2] / 2.0
+
+        # 1D equivalent -- out[-1] = (0 - y[-2])/2.0
+        slice1[axis] = -1
+        slice3[axis] = -2
+        out[slice1] = - f_data[slice3] / 2.0
+
+    # one-side differences
+    else:
+        # Numerical differentiation: 1st order edges
+        if f_data.shape[axis] == 2 or edge_order == 1:
+
+            slice1[axis] = 0
+            slice2[axis] = 1
+            slice3[axis] = 0
+            # 1D equivalent -- out[0] = (y[1] - y[0])
+            out[slice1] = (f_data[slice2] - f_data[slice3])
+
+            slice1[axis] = -1
+            slice2[axis] = -1
+            slice3[axis] = -2
+            # 1D equivalent -- out[-1] = (y[-1] - y[-2])
+            out[slice1] = (f_data[slice2] - f_data[slice3])
+
+        # Numerical differentiation: 2nd order edges
+        else:
+
+            slice1[axis] = 0
+            slice2[axis] = 0
+            slice3[axis] = 1
+            slice4[axis] = 2
+            # 1D equivalent -- out[0] = -(3*y[0] - 4*y[1] + y[2]) / 2.0
+            out[slice1] = -(3.0 * f_data[slice2] - 4.0 * f_data[slice3] +
+                            f_data[slice4]) / 2.0
+
+            slice1[axis] = -1
+            slice2[axis] = -1
+            slice3[axis] = -2
+            slice4[axis] = -3
+            # 1D equivalent -- out[-1] = (3*y[-1] - 4*y[-2] + y[-3]) / 2.0
+            out[slice1] = (3.0 * f_data[slice2] - 4.0 * f_data[slice3] +
+                           f_data[slice4]) / 2.0
+
+    # divide by step size
+    out /= dx
+
+    return out
+
+
+# noinspection PyAbstractClass
+class DiscreteGradient(Operator):
+    """Gradient operator for any number of dimensions `n`.
+
+    Calls helper function `discrete_part_deriv` to calculate each component
+    of the resulting product space vector. For the adjoint of the `Gradient`
+    operator to match the negative `DiscreteDivergence` operator requires
+    `zero_padding`.
+    """
+
+    def __init__(self, space, voxel_size=(1,), edge_order=2,
+                 zero_padding=True):
+        """Initialize a `DiscreteGradient` operator instance.
+
+        Zero padding is required for the negative `DiscreteDivergence`
+        operator to match the adjoint of the `DiscreteGradient` operator.
+
+        Parameters
+        ----------
+        space : `FunctionSpace`
+            The space of elements which the operator is acting on
+        voxel_size : `n`-tuple of `floats`
+            `n`-tuple of scalars specifying the sample distances in each
+            dimension. Default distance is 1 in each direction
+        edge_order : {1, 2}, optional
+            Partial derivative is calculated using Nth order accurate
+            differences at the boundaries. Default edge order: 2
+        zero_padding : `bool`, optional
+            Implicit zero padding. Assumes values outside of `space` to be
+            zero. Default: True
+        """
+        self.voxel_size = voxel_size
```

Comment:
We should verify some thins here, type of space, length of voxel_size, edge order = 1 or 2